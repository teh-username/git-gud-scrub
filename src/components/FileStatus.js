import React from 'react';

export const getFilesByState = (fileStates, state, targetValue) =>
  Object.entries(fileStates).reduce(
    (acc, [fileName, fileState]) =>
      fileState[state] === targetValue ? [...acc, fileName] : acc,
    []
  );

export const FileStatusRow = ({ prefix, files }) =>
  files.map(fileName => <div key={`${prefix}-${fileName}`}>{fileName}</div>);

export const FileStatusTable = ({ fileStates }) => (
  <table className="table table-bordered">
    <thead>
      <tr>
        <th>Untracked</th>
        <th>Unmodified</th>
        <th>Modified</th>
        <th>Staged</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <FileStatusRow
            prefix="ut"
            files={getFilesByState(fileStates, 'tracked', false)}
          />
        </td>
        <td>
          <FileStatusRow
            prefix="umod"
            files={getFilesByState(fileStates, 'modified', false)}
          />
        </td>
        <td>
          <FileStatusRow
            prefix="mod"
            files={getFilesByState(fileStates, 'modified', true)}
          />
        </td>
        <td>
          <FileStatusRow
            prefix="st"
            files={getFilesByState(fileStates, 'staged', true)}
          />
        </td>
      </tr>
    </tbody>
  </table>
);

const FileStatus = ({ fileStates }) => (
  <div className="col-lg-9 col-lg-pull-9 col-sm-12">
    <h4>File Status</h4>
    <FileStatusTable fileStates={fileStates} />
  </div>
);

export default FileStatus;
