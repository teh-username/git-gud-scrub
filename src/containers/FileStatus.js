import { connect } from 'react-redux';

import FileStatus from '../components/FileStatus';
import { getFileStatus } from '../redux/modules/fileSystem';

const mapStateToProps = state => ({
  fileStates: getFileStatus(state),
});

export default connect(mapStateToProps, null)(FileStatus);
