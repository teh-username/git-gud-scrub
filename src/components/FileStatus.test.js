import React from 'react';
import FileStatus, {
  FileStatusRow,
  FileStatusTable,
  getFilesByState,
} from './FileStatus';

describe('FileStatus Component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<FileStatus fileStates={{}} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('FileStatusRow Component', () => {
    it('renders the correct amount of divs', () => {
      const wrapper = shallow(
        <FileStatusRow prefix="ut" files={['heyo', 'woo', 'woof']} />
      );
      expect(wrapper.at(0).find('div').length).toEqual(3);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('FileStatusTable Component', () => {
    it('renders the table correclty', () => {
      const wrapper = shallow(
        <FileStatusTable
          fileStates={{
            revan: {
              tracked: false,
              modified: undefined,
              staged: undefined,
            },
            malak: {
              tracked: true,
              modified: true,
              staged: true,
            },
            bastila: {
              tracked: true,
              modified: false,
              staged: true,
            },
          }}
        />
      );
      expect(wrapper.find({ prefix: 'ut' }).props()).toEqual({
        prefix: 'ut',
        files: ['revan'],
      });
      expect(wrapper.find({ prefix: 'mod' }).props()).toEqual({
        prefix: 'mod',
        files: ['malak'],
      });
      expect(wrapper.find({ prefix: 'umod' }).props()).toEqual({
        prefix: 'umod',
        files: ['bastila'],
      });
      expect(wrapper.find({ prefix: 'st' }).props()).toEqual({
        prefix: 'st',
        files: ['malak', 'bastila'],
      });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('getFilesByState function', () => {
    it('builds the correct array', () => {
      const fileStates = {
        revan: {
          tracked: false,
          modified: undefined,
          staged: undefined,
        },
        malak: {
          tracked: true,
          modified: true,
          staged: true,
        },
        bastila: {
          tracked: true,
          modified: false,
          staged: true,
        },
      };
      expect(getFilesByState(fileStates, 'tracked', false)).toEqual(['revan']);
      expect(getFilesByState(fileStates, 'tracked', true)).toEqual([
        'malak',
        'bastila',
      ]);
      expect(getFilesByState(fileStates, 'modified', true)).toEqual(['malak']);
      expect(getFilesByState(fileStates, 'staged', true)).toEqual([
        'malak',
        'bastila',
      ]);
    });
  });
});
