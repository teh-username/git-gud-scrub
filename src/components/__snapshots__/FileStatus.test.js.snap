// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileStatus Component FileStatusRow Component renders the correct amount of divs 1`] = `
Array [
  <div
    key="ut-heyo"
/>,
  <div
    key="ut-woo"
/>,
  <div
    key="ut-woof"
/>,
]
`;

exports[`FileStatus Component FileStatusTable Component renders the table correclty 1`] = `
<table
  className="table table-bordered"
>
  <thead>
    <tr>
      <th>
        Untracked
      </th>
      <th>
        Unmodified
      </th>
      <th>
        Modified
      </th>
      <th>
        Staged
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Component
          files={
            Array [
              "revan",
            ]
          }
          prefix="ut"
        />
      </td>
      <td>
        <Component
          files={
            Array [
              "bastila",
            ]
          }
          prefix="umod"
        />
      </td>
      <td>
        <Component
          files={
            Array [
              "malak",
            ]
          }
          prefix="mod"
        />
      </td>
      <td>
        <Component
          files={
            Array [
              "malak",
              "bastila",
            ]
          }
          prefix="st"
        />
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`FileStatus Component renders correctly 1`] = `
<div
  className="col-lg-9 col-lg-pull-9 col-sm-12"
>
  <h4>
    File Status
  </h4>
  <Component
    fileStates={Object {}}
  />
</div>
`;
